[![master](https://github.com/ladislav-hovan/sponge/actions/workflows/test.yaml/badge.svg?branch=main)](https://github.com/ladislav-hovan/sponge/actions/workflows/test.yaml)
[![devel](https://github.com/ladislav-hovan/sponge/actions/workflows/test.yaml/badge.svg?branch=devel)](https://github.com/ladislav-hovan/sponge/actions/workflows/test.yaml)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)


# SPONGE - Simple Prior Omics Network GEnerator
The SPONGE package generates human prior gene regulatory networks and
protein-protein interaction networks for the involved transcription
factors.


## Table of Contents
- [SPONGE - Simple Prior Omics Network GEnerator](#sponge---simple-prior-omics-network-generator)
  - [Table of Contents](#table-of-contents)
  - [General Information](#general-information)
  - [Features](#features)
  - [Setup](#setup)
  - [Usage](#usage)
    - [File formats](#file-formats)
    - [Container](#container)
  - [Project Status](#project-status)
  - [Room for Improvement](#room-for-improvement)
  - [Acknowledgements](#acknowledgements)
  - [Contact](#contact)
  - [License](#license)


## General Information
This repository contains the SPONGE package, which allows the generation
of human prior gene regulatory networks based mainly on the data from
the JASPAR database.
It also uses NCBI to find the human analogs of vertebrate transcription
factors, Ensembl to collect all the promoter regions in the human
genome, UniProt for symbol matching, and STRING to retrieve
protein-protein interactions between transcription factors.
Because it accesses these databases on the fly, it requires internet
access.

Prior gene regulatory networks are useful mainly as an input for tools
that incorporate additional sources of information to refine them.
The prior networks generated by SPONGE are designed to be compatible
with PANDA and related [NetZoo](https://github.com/netZoo/netZooPy)
tools.

The purpose of this project is to give the ability to generate prior
gene regulatory networks to people who do not have the knowledge or
inclination to do the genome-wide motif search, but would still like
to change some parameters that were used to generate publicly available
prior gene regulatory networks.
It is also designed to facilitate the inclusion of new information from
database updates into the prior networks.

If you just want to use the prior networks generated by the stable
version of SPONGE with the default settings, they are available on
[Zenodo](https://zenodo.org/records/13628784).


## Features
The features already available are:
- Generation of prior gene regulatory network
- Generation of prior protein-protein interaction network for
  transcription factors
- Automatic download of required files during setup
- Parallelised motif filtering
- Command line interface


## Setup
The requirements are provided in a `requirements.txt` file.

SPONGE can be installed via pip:

``` bash
pip install netzoopy-sponge
```

Alternatively, it can be installed by downloading this repository and
then installing with pip (possibly in interactive mode):

``` bash
git clone https://github.com/ladislav-hovan/sponge.git
cd sponge
pip install -e .
```


## Usage
SPONGE comes with a `netzoopy-sponge` command line script:

``` bash
# Get information about the available options
netzoopy-sponge --help
# Run the pipeline
netzoopy-sponge
```

The script comes with a lot of options, but the defaults are designed
to be sensible and the users do not have to change any of them unless
desired.

Within Python, the default workflow can be invoked as follows:

``` python
# Import the class definition
from sponge.sponge import Sponge
# Run the default workflow
sponge_obj = Sponge(run_default=True)
```

Much like the command line script, the Sponge class implements many
variables that give control over the process, and they can be changed
from their defaults.
For more information, you can run `help(Sponge)` after the import.

In case one needs more control over the individual steps, the workflow
in Python would be as follows:

``` python
# Import the class definition
from sponge.sponge import Sponge
# Create the SPONGE object
sponge_obj = Sponge()
# Select the vertebrate transcription factors from JASPAR
sponge_obj.select_tfs()
# Find human homologs for the TFs if possible
sponge_obj.find_human_homologs()
# Filter the matches of the JASPAR bigbed file to the ones in the
# promoters of human transcripts
sponge_obj.filter_matches()
# Aggregate the filtered matches on promoters to genes
sponge_obj.aggregate_matches()
# Write the final motif prior to a file
sponge_obj.write_motif_prior()
# Retrieve the protein-protein interactions between the transcription
# factors from the STRING database
sponge_obj.retrieve_ppi()
# Write the PPI prior to a file
sponge_obj.write_ppi_prior()
```

SPONGE will attempt to download the files it needs into a temporary
directory (`.sponge_temp` by default).
Paths can be provided if these files were downloaded in advance.
The JASPAR bigbed file required for filtering is huge (> 100 GB), so
the download might take some time.
Make sure you're running SPONGE somewhere that has enough space!

As an alternative to the bigbed file download, SPONGE can download
tracks for individual TFs on the fly and filter them individually.
This way of processing is slower than the bigbed file when all TFs in
the database are considered, but it becomes competitive when only
a subset is used.
The physical storage footprint is much reduced.
The option is enabled with `on_the_fly_processing=True`.


### File formats
Users are free to provide their own files for the list of regions of
interest (key name `promoter`, default name `promoters.bed`), mapping
of transcripts to genes (`ensembl`: `ensembl.tsv`) and the list of
predicted TF binding sites (`jaspar_bigbed`: `JASPAR.bb`).
By default, if the paths are not provided through the keyword
`paths_to_files`, SPONGE attempts to locate these files in the temporary
folder under the default names.
If it fails to do so, it will proceed to download them.

List of regions of interest expects a bed file in the 6 column format
without a header, for example:

```
chr1    11119   12119   ENST00000456328 0       +
chr1    11260   12260   ENST00000450305 0       +
chr1    17186   18186   ENST00000619216 0       -
chr1    24636   25636   ENST00000488147 0       -
```

Mapping of transcripts to genes expects a four column tsv file with
a defined header, as an example:

```
Transcript stable ID    Gene stable ID  Gene name       Gene type
ENST00000387314 ENSG00000210049 MT-TF   Mt_tRNA
ENST00000389680 ENSG00000211459 MT-RNR1 Mt_rRNA
ENST00000387342 ENSG00000210077 MT-TV   Mt_tRNA
ENST00000387347 ENSG00000210082 MT-RNR2 Mt_rRNA
ENST00000386347 ENSG00000209082 MT-TL1  Mt_tRNA
ENST00000361390 ENSG00000198888 MT-ND1  protein_coding
```

The `Transcript stable ID` field will be used to match regions of
interest.
Finally, the predicted TF binding sites are expected in a binary bigbed
file, with the following format when decoded:

```
chrom   start      end      name  score strand TFName
chr1   10000    10006  MA0467.3    276      -    Crx
chr1   10000    10006  MA0648.2    233      +    GSC
chr1   10000    10006  MA0682.3    231      +  PITX1
chr1   10000    10006  MA0711.2    198      +   OTX1
chr1   10000    10006  MA0714.2    246      +  PITX3
```

Effectively, it is an extended bed format with a header, which uses
the `name` column to provide JASPAR matrix ID and the `TFName` column
to provide the actual name of the transcription factor.
However, currently SPONGE expects a bigbed file and will not work with
a bed file.


### Container
SPONGE releases are also provided as Docker containers.
The most basic way of running would involve mounting a directory to the
`/data` directory on the container, where networks will be written by
default:

``` bash
docker run --mount type=bind,source="$(pwd)"/output,target=/data ghcr.io/ladislav-hovan/netzoopy_sponge:latest
```

Help can be requested with the `--help` argument.
For the most part the arguments match those of the `netzoopy-sponge`
command line script, but interactive prompts are disabled.


## Project Status
The project is: _in progress_.


## Room for Improvement
Room for improvement:
- Try incorporating unipressed
- Improve overlap computations

To do:
- Support for more species


## Acknowledgements
Many thanks to the members of the
[Kuijjer group](https://www.kuijjerlab.org/)
at NCMM for their feedback and support.

This README is based on a template made by
[@flynerdpl](https://www.flynerd.pl/).


## Contact
Created by Ladislav Hovan (ladislav.hovan@ncmm.uio.no).
Feel free to contact me!


## License
This project is open source and available under the
[GNU General Public License v3](LICENSE).